passes/techmap/clock_gate.cc

```diff
@@ -0,0 +1,205 @@
+#include "kernel/yosys.h"
+#include "kernel/ff.h"
+#include <optional>
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct ClockGateCell {
+	std::string name;
+	std::string ce_pin;
+	std::string clk_in_pin;
+	std::string clk_out_pin;
+
+	ClockGateCell(std::string& name, std::string& str) : name(RTLIL::escape_id(name)) {
+		char delimiter = ':';
+		size_t pos1 = str.find(delimiter);
+		if (pos1 == std::string::npos)
+			log_assert(false && "Not enough ports in descriptor string");
+		size_t pos2 = str.find(delimiter, pos1 + 1);
+		if (pos2 == std::string::npos)
+			log_assert(false && "Not enough ports in descriptor string");
+		size_t pos3 = str.find(delimiter, pos2 + 1);
+		if (pos3 != std::string::npos)
+			log_assert(false && "Too many ports in descriptor string");
+
+		ce_pin = str.substr(0, pos1);
+		ce_pin = RTLIL::escape_id(ce_pin);
+
+		clk_in_pin = str.substr(pos1 + 1, pos2 - (pos1 + 1));
+		clk_in_pin = RTLIL::escape_id(clk_in_pin);
+
+		clk_out_pin = str.substr(pos2 + 1, str.size() - (pos2 + 1));
+		clk_out_pin = RTLIL::escape_id(clk_out_pin);
+	}
+};
+
+struct ClockgatePass : public Pass {
+	ClockgatePass() : Pass("clock_gate", "extract clock gating out of flip flops") { }
+	void help() override {
+		// TODO
+	}
+
+	SigMap sigmap;
+	FfInitVals initvals;
+
+	// One ICG will be generated per ClkNetInfo
+	// if the number of FFs associated with it is sufficent
+	struct ClkNetInfo {
+		// Original, ungated clock into enabled FF
+		Wire* clk_net;
+		// Original clock enable into enabled FF
+		Wire* ce_net;
+		bool pol_clk;
+		bool pol_ce;
+		unsigned int hash() const {
```

Comment:
Can we build a tuple and defer to the implicit hashing on that?

Reply:
I don't like the result:
```cpp
auto t = std::make_tuple(clk_net, ce_net, pol_clk, pol_ce);
unsigned int h = mkhash_init;
h = mkhash(h, hash_ops<std::tuple<Yosys::RTLIL::Wire *, Yosys::RTLIL::Wire *, bool, bool>>::hash(t));
return h;
```

Reply:
I forgot I can just `h = mkhash(h, hash_ops<decltype(t)>::hash(t));`

Reply:
> hash_ops<decltype(t)>::hash(t)

Or just use the `mkhash` helper (there are two functions named `mkhash`)

I assume you are shaping this into a hash-in hash-out pattern for #4524, otherwise there's no reason to return the tuple hash directly.

Reply:
Yep this is temporary